{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"V_PBFP7XdV17"},"outputs":[],"source":[]},{"cell_type":"markdown","source":["### **BACKGROUND**\n","In this code, we will write a program to detect ORFs in a given input nucleotide sequence. We will start by writing an algorithm based on our understanding of ORFs.\n"],"metadata":{"id":"nFOKISn9hxMz"}},{"cell_type":"markdown","source":["# Final Algorithm\n","\n","\"CATGCATGCATGAGTCGATCAGTCAGTGAGCGATGCATCAGTC\"\n","\n","Steps\n","\n","1. Open the seq to be able to read.\n","2. number the nucleotides\n","3. define/open the universal codon table  for start and stop codon\n","4.  break the seq into pair of 3nts 6. 5. starting from nts 1\n","* make it frame 1\n","6. start matching with the start codon 8. data nt by nt\n","9. keep track of the nts number\n","10. if match found of start codon\n","11. record the position of nts number\n","13. start matching with the stop codon data\n","14. if match found for stop codon\n","record the nts number & ORF number\n","15. take the seq between start & stop number\n","16. repeat the step 6 to 9\n","17. if match not found for stop codon\n","move to next frame\n","12. else match not found move to next frame\n","18. break the seq into 3 from nts number 2\n","19. name this new frame 2\n","20. Repeat from 6 to 10 for frame 2\n","21. Repeat from 6 to 10 for frame 3"],"metadata":{"id":"ErFyn3fCAv3E"}},{"cell_type":"markdown","source":["# Pseudo Code\n"],"metadata":{"id":"ch8h6-P9bv5U"}},{"cell_type":"markdown","source":["# Code\n","## Defining variables\n","now we will define variable "],"metadata":{"id":"kPiQFnY2b1c_"}},{"cell_type":"code","source":["!pip install Biopython\n","\n","from Bio import SeqIO"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ysAncWJF05zc","executionInfo":{"status":"ok","timestamp":1678758675292,"user_tz":-330,"elapsed":7828,"user":{"displayName":"SALIM SHAIKH","userId":"05984701926290095753"}},"outputId":"df2c416f-d596-4a24-9d47-c7402b1fb7a3"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting Biopython\n","  Downloading biopython-1.81-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m33.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.9/dist-packages (from Biopython) (1.22.4)\n","Installing collected packages: Biopython\n","Successfully installed Biopython-1.81\n"]}]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"W0vE8XuD6LNc","executionInfo":{"status":"ok","timestamp":1678758707831,"user_tz":-330,"elapsed":24119,"user":{"displayName":"SALIM SHAIKH","userId":"05984701926290095753"}},"outputId":"ef93d597-e19a-4243-a6f9-bcea7b3ef371"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}]},{"cell_type":"code","source":["# Import the SeqIO module from Biopython and read the fasta file into the 'record' object\n","record = SeqIO.read(\"/content/drive/MyDrive/Life Sciences/phage_MS2_genome_sequence.fasta\", \"fasta\")\n","\n","# Extract the DNA sequence from the 'record' object and assign it to the 'input_seq' variable\n","input_seq = record.seq\n","\n","# Define the start codon sequence as \"ATG\" and the stop codon sequences as \"TGA\", \"TAG\", and \"TAA\"\n","start_codons_seq = \"ATG\"\n","stop_codons_seqs = [\"TGA\",\"TAG\",\"TAA\"]\n","\n","# Initialize variables for the ORF number, frame number, ORF begin position, ORF end position, and ORF sequences\n","ORF_number = 0\n","frame_number = 1\n","ORF_begin = 0\n","ORF_end = 0\n","ORF_seqs = dict()\n","\n","# Define the length of ORFs to be 30 nucleotides\n","length_value = 30\n","\n","#defining start postion of the necluitide\n","nuc_pos = 0\n","\n","# Loop through each nucleotide in the input DNA sequence, checking for start and stop codons\n","# while loop will search through the entire DNA sequence, one codon at a time, until the end of the sequence is reached.\n","while nuc_pos < (len(input_seq) - 2):\n","  # This line extracts a three-nucleotide codon from the DNA sequence starting at the current position and assigns it to the 'codon' variable.\n","  codon = (input_seq[nuc_pos:(nuc_pos + 3)])\n","  # If a start codon is found, loop through the sequence from that point on, checking for a stop codon\n","  if codon == start_codons_seq:\n","    # This line starts a for loop that will search through the DNA sequence starting at the position of the start codon, looking for the next stop codon.\n","    for nuc_pos_2 in range(nuc_pos + 3,(len(input_seq) - 2),3):\n","      # This line extracts a three-nucleotide codon from the DNA sequence starting at the current position in the for loop and assigns it to the 'codon1' variable.\n","      codon1 = (input_seq[nuc_pos_2:(nuc_pos_2 + 3)])\n","      # This line checks if the 'codon1' variable matches any of the stop codon sequences specified earlier. If it does, the code will check if the length of the ORF is greater than the minimum length specified earlier. If it is, the ORF is considered a valid hit and its details are recorded.\n","      if codon1 in stop_codons_seqs:\n","        # If a stop codon is found, and the length of the ORF is greater than 30 nucleotides, add it to the ORF_seqs dictionary\n","        # These lines increase the ORF count, assign the beginning and end positions of the ORF, store the ORF sequence in the 'ORF_seqs' dictionary, and move the search position to the end of the ORF so that the code will not search within it again.\n","        if (nuc_pos_2 - nuc_pos) > length_value:\n","          ORF_number = ORF_number + 1\n","          ORF_begin = nuc_pos\n","          ORF_end = nuc_pos_2 + 2\n","          ORF_seqs[\"ORF\" + str(ORF_number)] = input_seq[nuc_pos:nuc_pos_2 + 3]\n","          nuc_pos = nuc_pos_2\n","          # This line exits the for loop once a stop codon is found.\n","        break\n","        # Continue searching for ORFs until the end of the sequence is reached\n","        # This line increments the search position by three nucleotides to move to the next codon.\n","  nuc_pos = nuc_pos + 3\n","    \n","# Print the number of ORFs, ORF begin position, ORF end position, and ORF sequences\n","print(ORF_number)\n","print(ORF_begin)\n","print(ORF_end)\n","print(ORF_seqs)\n"],"metadata":{"id":"pPUIwF9-cdex","executionInfo":{"status":"ok","timestamp":1678758713696,"user_tz":-330,"elapsed":363,"user":{"displayName":"SALIM SHAIKH","userId":"05984701926290095753"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"16b674f8-94b7-4eb8-99c1-e0ee2a730b1e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["7\n","2922\n","2970\n","{'ORF1': Seq('ATGCCATTTTTAATGTCTTTAGCGAGACGCTACCATGGCTATCGCTGTAGGTAG'), 'ORF2': Seq('ATGAGTGATATCCAGGGTGCATATGAGATGCTTACGAAGGTTCACCTTCAAGAG...TAG'), 'ORF3': Seq('ATGGAAACCCGATTCCCTCAGCAATCGCAGCAAACTCCGGCATCTACTAATAGA...TAA'), 'ORF4': Seq('ATGACAAATCCTTGTCATGGGATCCGGATGTTTTACAAACCAGCATCCGTAGCC...TGA'), 'ORF5': Seq('ATGATCAATCGATCAACCAGCGTCTGGCTCAGCAGGGCAGCGTAG'), 'ORF6': Seq('ATGGCGAGACGATACGATGGGAACTATTTTCCACAATGGGAAATGGGTTCACAT...TAG'), 'ORF7': Seq('ATGTCAAACCGTTTTACATCAAGAAACCTGTTGACAATCTCTTCGCCCTGA')}\n"]}]},{"cell_type":"markdown","source":["# Algorithm 1\n","Input: Neculotide sequence\n","\n","Output: ORF Found Sequence.\n","\n","- Took a neculotide sequence of Nipah Virus complete genome.\n","- Break down the neculotide sequence into 3 codons.\n","- look into the chart of start codon and stop codon sequence.\n","- looking for the start codon(ATG) in the given sequence.\n","- start codon found, start making ORF\n","- looking for stop codon(TAA, TAG, or TGA)\n","- stop codon found, stop making the ORF.\n","- write the ORF sequence found."],"metadata":{"id":"EtjIeP51nvSV"}},{"cell_type":"markdown","source":["# Algorithm 2\n","- input a neculotide sequence\n","- create a list of it and save it in a file\n","- break the neculotide sequence into 3 codons\n","- create a file to save the ORF\n","- read the ORF file created\n","- find start codon in the Sequence.(ATG)\n","- if not found go to the next frame \n","- if found, search for stop codon.(TAA, TAG, or TGA)\n","- if stop codon found, stop coding and move to find another start codon\n","- in loop run until the gemone end sequnce\n","- show the ORFs found in the result."],"metadata":{"id":"juzwKdYoiHMe"}}]}