{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyNpe1Z+Wc+gfWquEUQsI9GZ"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Primer Desingner\n","## Formula for temperature ratio for primers making.\n","45% > ratio of G+C / ratio of A+T+G+C < 55%\n","\n","1. input: 5 prime to 3 prime. \n","2. range/boundary of the selected gene.\n","3. extra nedded base pair nearby the range or Amplied region.\n","\n","\n"],"metadata":{"id":"wBzq4aPuu-3B"}},{"cell_type":"markdown","source":["#Algroithem\n","Input sequence: \n","\n","5' CGACCCGGAAAGGTTGCACAACTTAGACCCTACAATGTACTAGTAGGCCT 3'\n","\n","1. take sequence\n","2. start a count from 1 to 30 base.\n","3. look for the Ratio of A, T, G, Cs from 1 to 30.\n","3. from 30 base to 80 base is the amlified region.\n","4. look for the A, T, G, Cs from 80 to ...\n","5. \n","2. start to read how many A, T, G, Cs are there in the given sequence\n","3. then put the values got into the formula of ratio, to decide the temperature.\n","4. "],"metadata":{"id":"vp0uUzf5CpTW"}},{"cell_type":"markdown","source":["# Algrothem for forward primer\n","0.1. I/P sequence. \n","0.2. ask for range to be amplified.\n","0.3. max buffer = 200 ; length of primer = 19.\n","1. read sequence start to start + 19\n","2. find A, T, G, Cs ratio.\n","3. check if G/E % is alloweded range. \n","  - O/P seq, s , GC %\n","  - store \n","4. (s-1) to (s-1+19)\n","5. repeat step 2 and 3.\n","6. if reached s-200, break.\n","7. sort according to GC% distance from 50%.\n","# Algorithem for reverse sequence\n","0.5 take complenemtary sequence \n","1. read complemantary sequence from (e-19)-e\n","2. same as forward prime \n","3. same \n","  - store e -19 , reverse sequence, GC%\n","4. (e-19+1) to (e+1)\n","5. repeat step 2 and 3.\n","6. if reached (e+200-20)"],"metadata":{"id":"F42X0aTMMHd8"}},{"cell_type":"code","source":["# defining a function for calculating GC percent\n","def calcGC(my_seq):\n","  # counting every neculiotide\n","  g_count = my_seq.count('G')\n","  c_count = my_seq.count('C')\n","  # formula for counting GC% in the given sequence\n","  GC_per = 100 * (g_count + c_count)/ len(my_seq)\n","  return(GC_per) "],"metadata":{"id":"zQjb94s6-FCJ","executionInfo":{"status":"ok","timestamp":1678887789910,"user_tz":-330,"elapsed":4,"user":{"displayName":"SALIM SHAIKH","userId":"05984701926290095753"}}},"execution_count":2,"outputs":[]},{"cell_type":"code","execution_count":3,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Cyxgvy_MkSzQ","executionInfo":{"status":"ok","timestamp":1678887798410,"user_tz":-330,"elapsed":446,"user":{"displayName":"SALIM SHAIKH","userId":"05984701926290095753"}},"outputId":"559a3c9c-378e-406f-ef8e-5748500163c3"},"outputs":[{"output_type":"stream","name":"stdout","text":["50.0\n","ATGTACGACCCGGAAAGGTT\n","50.0\n","AATGTACGACCCGGAAAGGT\n","50.0\n","ACAATGTACGACCCGGAAAG\n"]}],"source":["# Set input sequence as a string\n","input_sequence = \"ATAAGACTCCTTAAGTTCTCTACAATGTACGACCCGGAAAGGTTGCACAACTTAGACCCTACAATGTACTAGTAGGCCTCTGCGCTGGCAATACAGATAAGATAATGTAGCCCCTGGCCTCAAAGGAACTCTCCTCCTTAGGTTGCATTTGTATAATGTTTGATTTTTAGAACAAAGGTTGCATGTACTAGTTCTCCTCCTTAAGTTCTCGTTCTCCTCCAATAAGAATAAGATAATGTT\"\n","\n","# Set start and stop regions of the forward primer and the maximum buffer\n","start_region = 29\n","stop_region = 79\n","max_buffer = 10\n","\n","# Set primer length\n","primer_length = 20\n","\n","# For loop to iterate through primer start positions\n","for primer_start in range(start_region, (start_region - max_buffer), -1): \n","\n","  # Read input sequence from primer_start to primer_start+primer_length\n","  read_sequence = (input_sequence[primer_start:(primer_start + primer_length)])\n","  \n","  # Calculate the ratio of GC to AT nucleotides in the read sequence\n","  if 45 < calcGC(read_sequence) < 55:\n","    # Set the output sequence to be the read sequence\n","    output_seq = read_sequence\n","    \n","    # Print the GC ratio and the output sequence\n","    print(calcGC(read_sequence))\n","    print(output_seq)\n"]}]}